imbi/automations.py

```diff
@@ -1,136 +1,350 @@
 import asyncio
 import dataclasses
 import logging
+import pathlib
+import tempfile
 import typing
 
+import cookiecutter.main
+import isort.api
 import sprockets_postgres
+import yarl
+from yapf.yapflib import yapf_api
 
 import imbi.integrations
 import imbi.user
 from imbi.endpoints import gitlab
 
 
+@dataclasses.dataclass
+class CookieCutter:
+    name: str
+    project_type_id: int
+    url: str
+
+
+@dataclasses.dataclass
+class Namespace:
+    gitlab_group_name: typing.Union[str, None]
+    name: str
+    slug: str
+
+
+@dataclasses.dataclass
+class ProjectType:
+    id: int
+    environment_urls: bool
+    gitlab_project_prefix: typing.Union[str, None]
+    name: str
+    slug: str
+
+
 @dataclasses.dataclass
 class Project:
     description: typing.Union[str, None]
     gitlab_project_id: typing.Union[int, None]
+    id: int
     name: str
-    namespace_id: int
-    project_id: int
-    project_type_id: int
+    namespace: Namespace
+    project_type: ProjectType
     slug: str
 
-    @classmethod
-    def from_database(cls, row: dict) -> 'Project':
-        field_names = {f.name for f in dataclasses.fields(cls)}
-        filtered = {n: v for n, v in row.items() if n in field_names}
-        filtered['project_id'] = row['id']
-        return cls(**filtered)
-
 
-class GitLabCreateProjectAutomation:
+class Automation:
     def __init__(self,
-                 automation_settings: dict,
-                 project: Project,
+                 application: 'imbi.app.Application',
                  user: 'imbi.user.User',
                  db: sprockets_postgres.PostgresConnector):
+        self.application = application
+        self.automation_settings = self.application.settings['automations']
         self.db = db
-        self.imbi_project = project
+        self.errors: typing.List[str] = []
         self.logger = logging.getLogger(__package__).getChild(
             self.__class__.__name__)
-        self.settings = automation_settings.get('gitlab', {})
         self.user = user
 
-        self._errors: typing.List[str] = []
+    def _add_error(self, msg_format, *args):
+        message = msg_format.format(*args)
+        self.logger.warning('%s', message)
+        self.errors.append(message)
+
+    def _has_error(self) -> bool:
+        return len(self.errors) > 0
+
+    async def _get_project(self, project_id: int) -> typing.Union[
```

Comment:
I believe the pythonic `typing` class here is `Optional`

Reply:
I'm on the fence with that interpretation.  I prefer the explicitness of `Union[T, None]`.  I question the usefulness of `typing.Optional` since it seldom means this _value is optional_.  For example the `print` function would be typed as:
```python
def print(value: typing.Any, *values: typing.Any, sep: str = ' ', end: str = '\n',
          file: io.IOBase = sys.stdout, flush: bool = False):
    ....
```

It would be awesome to say `file: typing.Optional[io.IOBase] = sys.stdout` since `file` is a _optional_ parameter but we cannot since it is not allowed to be `None`.  I might have to dig into the PEPs and see if I can convince my brain to accept the usage but I find `typing.Optional` to be conceptually more trouble than it is worth.